### PR TITLE
GameSA inline assembly cleanup

### PR DESCRIPTION
**Pull request by Merlin**
Removes most of the inline assembly for GameSA.
Only hooks and some function calls that are patched or unused are left.

I looked over all the files. But I may have messed up somewhere and didn't spot it.

Next thing I would like to do is cleaning up all the includes. 
And move all the *SAInterface classes to their own files and remove unused classes like CFont and a bunch of unused methods.

But before I do any of that it would be nice to know if these changes would be merged.

### CODE REVIEW

**Comment by patrikjuvonen** [comment]
Any changes this big should be made in small chunks that can be reviewed by anyone, merged into a branch one by one, and then the branch can be merged into master when complete.

Now there is a lot to review at once.

We can't risk "I might have missed something" in these large pull requests.

**Comment by sbx320** [comment]
It doesn't really matter too much whether this is a single PR with many commits or many PRs with many commits. As the changes are very localized (one isolated change per function), it's actually fairly easy to review in bulk. Just a lot of work to do. 

I've thus far reviewed it up to CAutomobileSA.h and it looks good, might get to some more over the weekend. Merging however will be a bit more annoying, since we're unlikely to be able to test it entirely before that. Guess we'll just have to time it with a stable build and risk a broken nightly.

File: Client/game_sa/CCameraSA.cpp
Comment (sbx320) [inline Client/game_sa/CCameraSA.cpp:320]: This doesn't look correct. 

If we aren't using it, we might as well remove it entirely.

File: Client/game_sa/CEntitySA.cpp
Comment (sbx320) [inline Client/game_sa/CEntitySA.cpp]:
```suggestion
    return ((CAnimBlendAssociation*(__cdecl*)(RpClump*, const char*))FUNC_RpAnimBlendClumpGetAssociation)(m_pInterface->m_pRwObject, szAnimName) != nullptr;
```

**Comment by Merlin** [comment]
Sorry, I thought it would be possible to review and approve single commits from one PR till all are approved and then merge it. Never done any code reviews on github. Should i still split it? And if, how would I do that without creating a new branch and PR?

File: Client/game_sa/CAutomobileSA.cpp
Comment (sbx320) [inline Client/game_sa/CAutomobileSA.cpp:79]: Random questionmark?

File: Client/game_sa/CGameSA.cpp
Comment (sbx320) [inline Client/game_sa/CGameSA.cpp:444]: Might be better to use CLASS_RwCamera here rather than the raw address

File: Client/game_sa/CHudSA.cpp
Comment (sbx320) [inline Client/game_sa/CHudSA.cpp:174]: I'm slightly confused on what the purpose of this is...

Good thing it's gone now

File: Client/game_sa/CHudSA.cpp
Comment (sbx320) [inline Client/game_sa/CHudSA.cpp]: `lea eax, dwColor`
`push eax`
```suggestion
    ((void(__cdecl*)(float, float, float, float, float, float, float, float, DWORD*))FUNC_Draw2DPolygon)(fX1, fY1, fX2, fY2, fX3, fY3, fX4, fY4, &dwColor);
```

File: Client/game_sa/CObjectSA.cpp
Comment (sbx320) [inline Client/game_sa/CObjectSA.cpp]: In the asm there's an additional push 1. Should be checked.

File: Client/game_sa/CPathFindSA.cpp
Comment (sbx320) [inline Client/game_sa/CPathFindSA.cpp]: `lea eax, nodeptr`
`push eax`
```suggestion
         FUNC_FindNthNodeClosestToCoors)(CLASS_CPathFind, &nodeptr, *vecCoors, 0, fDistance, false, false, iNodeNumber, iType, false, &node);
```

**Comment by sbx320** [comment]
There really isn't a great way to do large scale PRs via Github. You'd need to create a PR for each commit for us to approve & merge individually. Doesn't really scale for this amount of commits though.

However we can just keep this in here and merge manually via git itself if we need to partially merge.

--- 

Generally I'm tending towards just trying to get some kind of nightly test build going with these changes merged (not the "default" nightly build though). Most issues that would be caused by these changes should result in immediate crashes, so it would be good to try it out on some common servers. 

Anyway, 50/76 files reviewed thus far. Can't submit it until I'm actually done though...

File: Client/game_sa/CPlayerInfoSA.cpp
Comment (sbx320) [inline Client/game_sa/CPlayerInfoSA.cpp:109]: The function call is missing here

File: Client/game_sa/CPlayerInfoSA.cpp
Comment (sbx320) [inline Client/game_sa/CPlayerInfoSA.cpp:167]: the push ecx seems weird, need to double check this

File: Client/game_sa/CRestartSA.cpp
Comment (sbx320) [inline Client/game_sa/CRestartSA.cpp:48]: Shouldn't the first CVector also be a pointer?

File: Client/game_sa/CRestartSA.cpp
Comment (sbx320) [inline Client/game_sa/CRestartSA.cpp:62]: Same here: first CVector should also be a pointer

File: Client/game_sa/CRestartSA.cpp
Comment (sbx320) [inline Client/game_sa/CRestartSA.cpp:76]: Where does the unsigned int come from?

File: Client/game_sa/CRestartSA.cpp
Comment (sbx320) [inline Client/game_sa/CRestartSA.cpp:90]: Same here: unsigned int?

File: Client/game_sa/CWorldSA.cpp
Comment (sbx320) [inline Client/game_sa/CWorldSA.cpp:203]: If the function isn't used, we should remove it entirely rather than leaving an empty shell behind

File: Client/game_sa/CWorldSA.cpp
Comment (sbx320) [inline Client/game_sa/CWorldSA.cpp:327]: This seems rather odd. 

Might be worth checking out why this calls the 3d variant.

File: Client/game_sa/CModelInfoSA.cpp
Comment (sbx320) [inline Client/game_sa/CModelInfoSA.cpp:308]: Missing implementation?

File: Client/game_sa/TaskAttackSA.cpp
Comment (sbx320) [inline Client/game_sa/TaskAttackSA.cpp:188]: If it's thiscall, shouldn't we be using thiscall?

**Review by sbx320 — CHANGES_REQUESTED** [review]
Found a few mistakes and unclear changes. Otherwise it looks good.

Long-term we might want to move the types into the address declarations, but that's a separate task.

File: Client/game_sa/CCameraSA.cpp
Comment (Merlin) [reply Client/game_sa/CCameraSA.cpp:320]: Unused and FUNC_TryToStartNewCamMode(0x467AAD) points to code inside of CRunningScript::ProcessCommands100To199. I will remove a bunch of unused code in a separate PR.

File: Client/game_sa/CAutomobileSA.cpp
Comment (Merlin) [reply Client/game_sa/CAutomobileSA.cpp:79]: CAutomobile::BlowUpCar has another parameter that was not pushed. It's also unused. Will also be removed in a separate PR

File: Client/game_sa/CPlayerInfoSA.cpp
Comment (Merlin) [reply Client/game_sa/CPlayerInfoSA.cpp:167]: CRemote::TakeRemoteControlledCarFromPlayer has only 1 parameter. Looks also to be unused

File: Client/game_sa/CRestartSA.cpp
Comment (Merlin) [reply Client/game_sa/CRestartSA.cpp:48]: No, first parameter is actually just a CVector. It's also unused

File: Client/game_sa/CRestartSA.cpp
Comment (Merlin) [reply Client/game_sa/CRestartSA.cpp:62]: Same as CRestart::FindClosestPoliceRestartPoint

File: Client/game_sa/CRestartSA.cpp
Comment (Merlin) [reply Client/game_sa/CRestartSA.cpp:90]: The unsigned int is the town where an restart point should be added. Also unused

File: Client/game_sa/CWorldSA.cpp
Comment (Merlin) [reply Client/game_sa/CWorldSA.cpp:203]: There will be an separate PR that removes unused stuff. Or do you want me to remove it in this PR?

File: Client/game_sa/CWorldSA.cpp
Comment (Merlin) [reply Client/game_sa/CWorldSA.cpp:327]: CWorldSA::FindGroundZForPosition is also never called and should be removed

File: Client/game_sa/CModelInfoSA.cpp
Comment (Merlin) [reply Client/game_sa/CModelInfoSA.cpp:308]: FUNC_GetLevelFromPosition(0x4DD300) points to code inside of CAEFireAudioEntity::PlayFireSound. I guess its left over from vice city days. Also unused. I just saw CWorldSA::GetLevelFromPosition is actually also broken and unused. Both should be removed

File: Client/game_sa/TaskAttackSA.cpp
Comment (Merlin) [reply Client/game_sa/TaskAttackSA.cpp:188]: CTaskSimpleUseGunSA::RequirePistolWhip is a static function and also unused

File: Client/game_sa/CRestartSA.cpp
Comment (Merlin) [reply Client/game_sa/CRestartSA.cpp:76]: The unsigned int is the town where an restart point should be added. Also unused

File: Client/game_sa/CGameSA.cpp
Comment (Merlin) [reply Client/game_sa/CGameSA.cpp:444]: CLASS_RwCamera points to TheCamera.m_pRwCamera. I think its cleaner the way it currently is. I can change it if you really want? It's also unused and should be removed

File: Client/game_sa/CWorldSA.cpp
Comment (sbx320) [reply Client/game_sa/CWorldSA.cpp:203]: A separate PR is fine, since it seems to be quite a few functions not being used.

File: Client/game_sa/CPlayerInfoSA.cpp
Comment (sbx320) [reply Client/game_sa/CPlayerInfoSA.cpp:167]: Yeah, reading into it, it seems like whoever wrote that asm just wanted to save the ecx register state.

File: Client/game_sa/CWorldSA.cpp
Comment (Merlin) [reply Client/game_sa/CWorldSA.cpp:203]: Should I base the PR on my inline_assembly_cleanup branch or multitheftauto:master?

File: Client/game_sa/CGameSA.cpp
Comment (sbx320) [reply Client/game_sa/CGameSA.cpp:444]: If it's unused it doesn't really matter anyway.

File: Client/game_sa/CWorldSA.cpp
Comment (sbx320) [reply Client/game_sa/CWorldSA.cpp:203]: Use your cleanup branch, otherwise we'll have merge conflicts everywhere when trying to merge both PRs

File: Client/game_sa/CRestartSA.cpp
Comment (sbx320) [reply Client/game_sa/CRestartSA.cpp:48]: Interesting. I suppose we got lucky and it never caused trouble since we didn't call it.

**Review by sbx320 — DISMISSED** [review]

**Comment by Pirulax** [comment]
`plugin-sdk` has a few functions meant for this exact purpose, those could've been used (`CallMethod`, etc..), it would've looked nicer, but still, good job!

**Review by sbx320 — APPROVED** [review]

**Comment by Merlin** [comment]
I think it will be easier if I slowly introduce these calls with further game sa cleanups.
Maybe with cleaned up game classes in separate files then their mirror/wrapper classes.
That way we can also include them in Deathmatch/MultiplayerSA for quick debugging without introducing a bunch of virtual functions in mirror/wrapper classes.
None of your review work would go to waste. I would just copy them into clean new classes.

**Comment by sbx320** [comment]
Yeah, resolving those conflicts might be painful. 

On a related note: I've been looking at in the past was to use some custom DSL tooling to automate manually writing these headers/calls. Essentially you'd create a file

```
class CVehicle 
{
     void BlowUp(bool bExplode) @ 0x123456; 
}
```

and it could generate the corresponding C++ code out of it. My initial work on that was using Xtext, but Java is annoying so I didn't get too far. Might be worth looking into again.

**Pull request closed without merging**
